File: tools/claude/config/skills/scan-git-for-tils/notion/commits.py
"""Notion assessed commits tracking utilities."""

from __future__ import annotations

from datetime import date

from notion.validation import NotionPageResponse
from op.secrets import OP_NOTION_TOKEN_PATH, get_op_secret

# Notion database IDs
ASSESSED_COMMITS_DATA_SOURCE_ID = "cba80148-aeef-49c9-ba45-5157668b17b3"
NOTION_ASSESSED_COMMITS_DB = "928fcd9e47a84f98824790ac5a6d37ca"


def get_assessed_commits_from_notion() -> set[str]:
    """Fetch all assessed commit hashes from Notion database."""
    from notion_client import Client
    from notion_client.helpers import collect_paginated_api

    token = get_op_secret(OP_NOTION_TOKEN_PATH)
    if not token:
        return set()

    try:
        notion = Client(auth=token)
    except Exception:
        return set()

    try:
        # Use helper to automatically handle pagination (Notion API v2025-09-03)
        pages = collect_paginated_api(
            notion.data_sources.query,
            data_source_id=ASSESSED_COMMITS_DATA_SOURCE_ID,
        )

        # Extract commit hashes from results
        assessed_hashes = set()
        for page in pages:
            title_prop = page.get("properties", {}).get("Commit Hash", {})
            title_content = title_prop.get("title", [])
            if title_content:
                commit_hash = title_content[0].get("plain_text", "")
                if commit_hash:
                    assessed_hashes.add(commit_hash)

        return assessed_hashes

    except Exception:
        return set()


def find_existing_tracker_entry(notion, commit_hash: str) -> str:
    """Check if tracker entry already exists for this commit. Returns page ID if found."""
    try:
        results = notion.data_sources.query(
            data_source_id=ASSESSED_COMMITS_DATA_SOURCE_ID,
            filter={"property": "Commit Hash", "title": {"equals": commit_hash}},
        )
        if results.get("results"):
            return results["results"][0]["id"]
    except Exception:
        pass

    return ""


def update_tracker_entry(notion, page_id: str, writing_page_id: str) -> str:
    """Update existing tracker entry to link to Writing page. Returns page URL."""
    try:
        response = notion.pages.update(
            page_id=page_id,
            properties={
                "Writing": {"relation": [{"id": writing_page_id}]},
                "Assessed": {"date": {"start": date.today().isoformat()}},
            },
        )
        # Parse response immediately to validate structure
        page = NotionPageResponse.model_validate(response)
        return page.url
    except Exception as e:
        raise Exception(f"Failed to update tracker: {e}")


def create_tracker_entry(notion, commit: dict, writing_page_id: str) -> str:
    """Create an entry in TIL Assessed Commits and link to Writing page. Returns page URL."""

    properties = {
        "Commit Hash": {"title": [{"type": "text", "text": {"content": commit["hash"]}}]},
        "Message": {"rich_text": [{"type": "text", "text": {"content": commit["message"][:2000]}}]},
        "Repo": {"rich_text": [{"type": "text", "text": {"content": commit["repo"]}}]},
        "Assessed": {"date": {"start": date.today().isoformat()}},
        "Writing": {"relation": [{"id": writing_page_id}]},
    }

    # Only add Commit Date if present (None breaks Notion API)
    if commit.get("date"):
        properties["Commit Date"] = {"date": {"start": commit["date"]}}

    response = notion.pages.create(
        parent={"data_source_id": ASSESSED_COMMITS_DATA_SOURCE_ID},
        properties=properties,
    )

    # Parse response immediately to validate structure
    page = NotionPageResponse.model_validate(response)
    return page.url

File: tools/claude/config/skills/scanning-git-for-tils/notion/commits.ts
/**
 * Notion assessed commits tracking - BUN VERSION
 * Notice: No Protocol hacks, no type: ignore comments
 */

import { Client } from "@notionhq/client@^2.2.15";
import { z } from "zod@^3.22.4";
import { getOpSecret, OP_NOTION_TOKEN } from "../op/secrets.ts";

const ASSESSED_COMMITS_DATA_SOURCE_ID = "cba80148-aeef-49c9-ba45-5157668b17b3";
const NOTION_ASSESSED_COMMITS_DB = "928fcd9e47a84f98824790ac5a6d37ca";

// Zod validates AND provides types - no separate validation needed
const NotionPageResponseSchema = z.object({
  url: z.string(),
  id: z.string().optional(),
});

const NotionDatabaseQueryResponseSchema = z.object({
  results: z.array(z.unknown()),
  has_more: z.boolean(),
  next_cursor: z.string().nullable(),
});

export async function getAssessedCommitsFromNotion(): Promise<Set<string>> {
  const token = await getOpSecret(OP_NOTION_TOKEN);
  if (!token) return new Set();

  try {
    const notion = new Client({ auth: token });
    const assessedHashes = new Set<string>();
    let startCursor: string | null = null;

    while (true) {
      const queryParams = startCursor
        ? { database_id: NOTION_ASSESSED_COMMITS_DB, start_cursor: startCursor }
        : { database_id: NOTION_ASSESSED_COMMITS_DB };

      const responseData = await notion.databases.query(queryParams);
      const response = NotionDatabaseQueryResponseSchema.parse(responseData);

      // Extract commit hashes
      for (const page of response.results) {
        if (typeof page !== "object" || !page) continue;
        const props = (page as Record<string, Record<string, unknown>>).properties;
        if (!props || typeof props !== "object") continue;

        const titleProp = props["Commit Hash"] as { title?: unknown[] };
        if (!titleProp?.title) continue;

        const titleContent = titleProp.title;
        if (!Array.isArray(titleContent) || titleContent.length === 0) continue;

        const commitHash = (titleContent[0] as { plain_text?: string })?.plain_text;
        if (typeof commitHash === "string" && commitHash) {
          assessedHashes.add(commitHash);
        }
      }

      if (!response.has_more) break;
      startCursor = response.next_cursor;
    }

    return assessedHashes;
  } catch {
    return new Set();
  }
}

export async function findExistingTrackerEntry(
  notion: Client,
  commitHash: string,
): Promise<string> {
  try {
    const responseData = await notion.databases.query({
      database_id: ASSESSED_COMMITS_DATA_SOURCE_ID,
      filter: { property: "Commit Hash", title: { equals: commitHash } },
    });

    const response = NotionDatabaseQueryResponseSchema.parse(responseData);
    if (response.results.length > 0) {
      const firstResult = response.results[0];
      if (typeof firstResult === "object" && firstResult && "id" in firstResult) {
        return String(firstResult.id);
      }
    }
  } catch {
    // Ignore errors
  }

  return "";
}

export async function updateTrackerEntry(
  notion: Client,
  pageId: string,
  writingPageId: string,
): Promise<string> {
  const pageData = await notion.pages.update({
    page_id: pageId,
    properties: {
      Writing: { relation: [{ id: writingPageId }] },
      Assessed: { date: { start: new Date().toISOString().slice(0, 10) } },
    },
  });

  const page = NotionPageResponseSchema.parse(pageData);
  return page.url;
}

export async function createTrackerEntry(
  notion: Client,
  commit: Record<string, string>,
  writingPageId: string,
): Promise<string> {
  const properties = {
    "Commit Hash": { title: [{ type: "text" as const, text: { content: commit.hash } }] },
    Message: {
      rich_text: [{ type: "text" as const, text: { content: commit.message.slice(0, 2000) } }],
    },
    Repo: { rich_text: [{ type: "text" as const, text: { content: commit.repo } }] },
    Assessed: { date: { start: new Date().toISOString().slice(0, 10) } },
    Writing: { relation: [{ id: writingPageId }] },
    ...(commit.date && { "Commit Date": { date: { start: commit.date } } }),
  };

  const pageData = await notion.pages.create({
    parent: { database_id: ASSESSED_COMMITS_DATA_SOURCE_ID },
    properties,
  });

  const page = NotionPageResponseSchema.parse(pageData);
  return page.url;
}
